File: scripts/backend-node/razorpay-migration/reconciliation.js
/**
 * Daily reconciliation cron stub.
 *
 * For each order created in the last 24h, pull final state from the gateway it
 * used and compare to your DB. Flag discrepancies.
 *
 * Replace `db.*` calls with your real persistence layer. The shape returned
 * here is what your monitoring / paging should consume.
 */

import Razorpay from "razorpay";
import PaytmChecksum from "paytmchecksum";

const RAZORPAY = new Razorpay({
  key_id: process.env.RAZORPAY_KEY_ID,
  key_secret: process.env.RAZORPAY_KEY_SECRET,
});
const PAYTM_BASE =
  (process.env.PAYTM_ENVIRONMENT || "staging") === "production"
    ? "https://secure.paytmpayments.com"
    : "https://securestage.paytmpayments.com";

async function fetchRazorpayStatus(razorpayOrderId) {
  const payments = await RAZORPAY.orders.fetchPayments(razorpayOrderId);
  // Pick latest captured payment, or the latest one if nothing captured.
  const captured = payments.items.find((p) => p.status === "captured");
  const latest = payments.items[0];
  return {
    found: !!(captured || latest),
    status: captured ? "TXN_SUCCESS" : latest?.status === "failed" ? "TXN_FAILURE" : "PENDING",
    raw: captured || latest,
  };
}

async function fetchPaytmStatus(orderId) {
  const body = { mid: process.env.PAYTM_MID, orderId };
  const signature = await PaytmChecksum.generateSignature(JSON.stringify(body), process.env.PAYTM_MERCHANT_KEY);
  const r = await fetch(`${PAYTM_BASE}/v3/order/status`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ head: { signature }, body }),
  });
  const json = await r.json();
  const status = json?.body?.resultInfo?.resultStatus || "UNKNOWN";
  return { found: status !== "UNKNOWN", status, raw: json?.body };
}

/**
 * Reconcile a single order. Returns { match: boolean, dbState, gatewayState }.
 */
export async function reconcileOrder(order) {
  const dbState = order.status; // your DB's view: TXN_SUCCESS / TXN_FAILURE / PENDING
  const gateway = order.psp === "razorpay"
    ? await fetchRazorpayStatus(order.pspOrderId)
    : await fetchPaytmStatus(order.orderId);

  return {
    orderId: order.orderId,
    psp: order.psp,
    dbState,
    gatewayState: gateway.status,
    match: dbState === gateway.status,
    found: gateway.found,
  };
}

/**
 * Run-once reconciliation. Wire to cron / cloud scheduler.
 */
export async function reconcileLast24h({ getOrders }) {
  const orders = await getOrders({ sinceHours: 24 });
  const results = await Promise.all(orders.map(reconcileOrder));

  const summary = {
    total: results.length,
    matches: results.filter((r) => r.match).length,
    mismatches: results.filter((r) => !r.match).length,
    byPsp: {
      razorpay: { total: 0, mismatches: 0 },
      paytm: { total: 0, mismatches: 0 },
    },
    mismatchDetail: results.filter((r) => !r.match),
  };
  for (const r of results) {
    summary.byPsp[r.psp].total++;
    if (!r.match) summary.byPsp[r.psp].mismatches++;
  }
  return summary;
}

/*
Example wire-up (cron every 2h or so):

import { reconcileLast24h } from "./reconciliation.js";
import { db } from "./db.js";

const summary = await reconcileLast24h({
  getOrders: ({ sinceHours }) => db.ordersCreatedSince(sinceHours),
});
if (summary.mismatches > 0) await alerts.page("PSP reconciliation drift", summary);
*/

File: scripts/backend-python/razorpay_migration/reconciliation.py
"""
Daily reconciliation cron stub.

For each order in the past 24h, fetch final state from whichever gateway it
used and compare to your DB. Flag drift; page above threshold.
"""

from __future__ import annotations

import json
import os
from dataclasses import dataclass
from typing import Callable, Iterable

import razorpay
import requests

from paytmchecksum import PaytmChecksum


def _paytm_base() -> str:
    return (
        "https://secure.paytmpayments.com"
        if os.getenv("PAYTM_ENVIRONMENT", "staging") == "production"
        else "https://securestage.paytmpayments.com"
    )


@dataclass
class Order:
    order_id: str
    psp: str  # "razorpay" | "paytm"
    psp_order_id: str
    status: str  # your DB's view


def _fetch_razorpay_status(razorpay_order_id: str) -> dict:
    client = razorpay.Client(auth=(os.getenv("RAZORPAY_KEY_ID"), os.getenv("RAZORPAY_KEY_SECRET")))
    payments = client.order.payments(razorpay_order_id)["items"]
    captured = next((p for p in payments if p["status"] == "captured"), None)
    latest = payments[0] if payments else None
    status = (
        "TXN_SUCCESS" if captured
        else "TXN_FAILURE" if (latest and latest["status"] == "failed")
        else "PENDING" if latest
        else "UNKNOWN"
    )
    return {"found": bool(captured or latest), "status": status, "raw": captured or latest}


def _fetch_paytm_status(order_id: str) -> dict:
    body = {"mid": os.getenv("PAYTM_MID"), "orderId": order_id}
    signature = PaytmChecksum.generateSignature(json.dumps(body), os.getenv("PAYTM_MERCHANT_KEY"))
    r = requests.post(
        f"{_paytm_base()}/v3/order/status",
        json={"head": {"signature": signature}, "body": body},
        timeout=15,
    )
    js = r.json()
    status = ((js.get("body") or {}).get("resultInfo") or {}).get("resultStatus", "UNKNOWN")
    return {"found": status != "UNKNOWN", "status": status, "raw": js.get("body")}


def reconcile_order(order: Order) -> dict:
    gateway = (
        _fetch_razorpay_status(order.psp_order_id)
        if order.psp == "razorpay"
        else _fetch_paytm_status(order.order_id)
    )
    return {
        "orderId": order.order_id,
        "psp": order.psp,
        "dbState": order.status,
        "gatewayState": gateway["status"],
        "match": order.status == gateway["status"],
        "found": gateway["found"],
    }


def reconcile_last_24h(get_orders: Callable[[int], Iterable[Order]]) -> dict:
    orders = list(get_orders(24))
    results = [reconcile_order(o) for o in orders]
    summary = {
        "total": len(results),
        "matches": sum(1 for r in results if r["match"]),
        "mismatches": sum(1 for r in results if not r["match"]),
        "byPsp": {
            "razorpay": {"total": 0, "mismatches": 0},
            "paytm": {"total": 0, "mismatches": 0},
        },
        "mismatchDetail": [r for r in results if not r["match"]],
    }
    for r in results:
        summary["byPsp"][r["psp"]]["total"] += 1
        if not r["match"]:
            summary["byPsp"][r["psp"]]["mismatches"] += 1
    return summary
